### ISSUE TITLE
Split up Arrow Crate

### ISSUE THREAD

**Issue by tustvold** [apache/arrow-rs#2594]
**TLDR rather than fighting entropy lets just brute-force compilation**

**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The arrow crate is getting rather large, and is starting to show up as a non-trivial bottleneck when compiling code, see #2170. There have been some efforts to reduce the amount of generated code, see #1858, but this is going to be a perpetual losing battle against new feature additions.

I think there are a couple of problems currently:

1. Limited build parallelism, especially if [codegen-units](https://doc.rust-lang.org/stable/rustc/codegen-options/#codegen-units) is set low
2. Upstream crates have to "depend" on functionality they don't need, e.g. `parquet` depending on compute kernels
3. Minor changes force large amounts of recompilation, with incremental compilation only helping marginally
4. Codegen is rarely linear in complexity, consequently larger codegen units take longer than the same amount of code in smaller units

All these conspire to often result in an `arrow` shaped hole in compilation, where CPUs are left idle.

Some numbers from my local machine

* Release with default features: 232 seconds
* Release with default features without comparison kernels: 150 seconds
* Release with default features without compute kernels: 70 seconds
* Release without default features without compute kernels: 60 seconds

**The vast majority of the time all bar a single core is idle.**

**Describe the solution you'd like**

I would like to propose we split up the arrow crate, into a number of sub-crates that are then re-exported by the top-level `arrow` crate. Users can then choose to depend on the batteries included `arrow` crate, or more granular crates.

Initially I would propose the following split:

* arrow-csv: CSV reader support
* arrow-ipc: IPC support
* arrow-json: JSON support (related to #2300)
* arrow-compute: contents of compute module
* arrow-test: arrow test_utils (not published)
* arrow-core: everything else

There is definitely scope for splitting up the crates further after this, in particular the comparison kernels might be a good candidate to live on their own, but I think lets start small and go from there. I suspect there is a fair amount of disentangling that will be necessary to achieve this.

**Describe alternatives you've considered**

Feature flags are another way this can be handled, however, they have a couple of limitations:

* It is impractical to test the full combinatorial explosion of combinations, which allows for bugs to sneak through
* They are unified for a target which limits build parallelism, just because say DataFusion depends on arrow with CSV support, shouldn't force the `parquet` crate to wait for this to compile before it can start compiling
* Poor UX:
  * Discoverability is limited, it can be hard to determine what features gate what functionality
  * Hard to determine if the feature flag set is minimal, no equivalent of cargo-udeps
  * It can be a non-trivial detective exercise to determine why a given feature is being enabled
  * Necessitate counter-intuitive hacks to play nicely in multi-crate workspaces - see [workspace hack](https://docs.rs/cargo-hakari/latest/cargo_hakari/about/index.html#what-are-workspace-hack-crates)

**Additional context**

@Jimexist recently drove an initiative to do something similar to DataFusion which has worked very well - https://github.com/apache/arrow-datafusion/issues/1750

FYI @alamb @jhorstmann @nevi-me

**Comment by tustvold** [comment]
I've started work on this with #2693, I think the final split will likely end up being a different from what was initially proposed based on what components can easily be separated. I next plan to split out array data, followed by the arrays themselves. This should then allow splitting out some of the heavier kernels, e.g. sort, compare, cast, etc...

**Comment by alamb** [comment]
If the history with the datafusion split is any indication, this work is likely to end up generating lots of PRs

You can see how @Jimexist broke that down and we tracked it in https://github.com/apache/arrow-datafusion/issues/1750 -- perhaps something similar could be applied here.

I am happy to try and review these mechanical PRs more quickly so we can get the project done more quickly

**Comment by andygrove** [comment]
How feasible would it be to move the type definitions (such as `DataType`, `Field`, and `Schema`) into an `arrow-types` crate? Maybe this is a path to `arrow2` being able to use a common type system.

**Comment by tustvold** [comment]
I don't see an obvious reason why that would not be possible, I'm not sure how generally useful the types will be without the array definitions though...

**Comment by andygrove** [comment]
The `datafusion-sql` crate only uses `arrow::datatypes` for example, so just depending on `arrow-types` there would presumably help with compilation times.

**Comment by jorgecarleitao** [comment]
Hey, Thanks for the ping!

I think it would not benefit arrow2 directly right now as it has different declarations for `Field` (e.g. we do not have dict_id on it). Arrow2 also has an extension (`DataType::Extension`).

With that said, imo it is still a good design - there are systems that only require `DataType`, `Field`, `Schema`, and functionality to read them from a file. One example is a data catalog based on arrow logical types.

I think that datafusion's logical plans could also only depend on types, but I could be wrong (it depends on how List scalars are represented there?).

**Comment by alamb** [comment]
> I think that datafusion's logical plans could also only depend on types, but I could be wrong (it depends on how List scalars are represented there?).

List scalars are represented as `Vec<Box<Field>` so I agree it may be the case the logical plans could depend only on type

https://github.com/apache/arrow-datafusion/blob/d16457a0ba129b077935078e5cf89d028f598e0b/datafusion/common/src/scalar.rs#L81

**Comment by tustvold** [comment]
I've created https://github.com/apache/arrow-rs/pull/2711 which splits out the schema definitions into a crate called arrow-schema. I thought this was more clearly the logical types than something called arrow-types. PTAL :smile:

**Comment by maxburke** [comment]
As a downstream user of Arrow, one of the things we find is that we need to fork Arrow-ecosystem crates to quickly integrate patches for missing features or bugs and I think one thing that I'm dreading is having to do that with an exploded Arrow crate, having to fork half a dozen Arrow packages, Parquet, all of the Datafusion crates, ..., seems like it'll be a royal pain.

**Comment by tustvold** [comment]
Hi @maxburke, the intention is to follow the work that was already performed for DataFusion, and would not involve splitting the repository. So I think you shouldn't have to maintain any more or less forks following this?

**Comment by alamb** [comment]
Put another way, @maxburke  -- if you use the `patch.crates-io` directive in your workspace `Cargo.toml` it will automatically pull in dependent crates

For example, to change datafusion to use arrow (and a new arrow-buffers) in https://github.com/apache/arrow-rs/pull/2693 I only had to add this to the `datafusion/Cargo.toml`

```toml
[patch.crates-io]
arrow = {  git = "https://github.com/tustvold/arrow-rs.git", rev="f47a878"}
parquet = {  git = "https://github.com/tustvold/arrow-rs.git", rev="f47a878" }
arrow-flight = {  git = "https://github.com/tustvold/arrow-rs.git", rev="f47a878" }
```

We have been following this model in IOx (where we use a different upstream datafusion). Specifically, adding these lines:

https://github.com/influxdata/influxdb_iox/blob/main/datafusion/Cargo.toml#L12-L14

```toml
upstream = { git = "https://github.com/apache/arrow-datafusion.git", rev="97b3a4b37f54aaa52f8705db3e57b15ee98c24a7", default-features = false, package = "datafusion" }
datafusion-proto = { git = "https://github.com/apache/arrow-datafusion.git", rev="97b3a4b37f54aaa52f8705db3e57b15ee98c24a7" }
```

Is enough to get us all the datafusion crates (like datafusion-core, datafusion-expr, etc....)

**Comment by alamb** [comment]
Maybe we could try to split one crate out in this next release and see how easy/hard it is on downstream consumers?

**Comment by alamb** [comment]
I wonder if we can claim this project is complete?

**Comment by tustvold** [comment]
Mostly, what do you think of splitting out the following and calling it a day

* arrow-row - row format
* arrow-ord - sort, comparison and partition kernels
* arrow-arith - arithmetic, boolean and temporal kernels

We can then move the window kernels to arrow-select, and all the kernels will have been moved out of the top-level crate.

I think that would be a concrete end milestone?

**Comment by alamb** [comment]
I wonder is it done yet 🙏

**Comment by tustvold** [comment]
arrow-row and arrow-arith then yes
